### DISCUSSION TITLE
Low recall while training NER pipeline

### DISCUSSION THREAD

**Discussion by Joozty** [explosion/spaCy#9379] [Help: Best practices]
I am training a NER pipeline. My datataset consist of Wikipedia pages and entities are internal links. I use config (optimised for accuracy) generated [here](https://spacy.io/usage/training).  Train dataset contains 50K articles (split by paragraphs) and dev dataset has 15K articles. There is almost 78K overlapping entities between train and dev.  Train dataset has 290K unique entities and dev has about 110K unique entities. Batch size is between 50-100 paragraphs.

My training is right now on 13th epoch and I get following results: Precision 76.68/Recall 61.92. and it's improving very slowly. So I wonder what parameters should I tweak? Do you have any recommendations? I am thinking about using [wand for searching right hyperparameters](https://github.com/explosion/projects/tree/v3/integrations/wandb) but its very costly for resources so I wanted ask you first maybe you can see some problem here. 🤔 

![image](https://user-images.githubusercontent.com/12775440/136163148-afaaf5bd-c81b-4f77-aebd-772f53bd37fe.png)

I should also mention that I use only one label for all entities. Is it a big problem? How many epochs you train your NER pipeline on?

Thanks in advance

**Comment by polm** [comment]
> My datataset consist of Wikipedia pages and entities are internal links.

Are you using all internal links in Wikipedia, not a subset? That's going to be hard because Wikipedia mentions don't always look like normal named entities; often they don't look like anything at all - you could have a sentence like `The Simpsons' house has a [[wood]] [[fence]] in the [[back yard]].`. 

To take a real example picked at random:

> Within the grounds are a bridge, a rotunda and a gatehouse, each of which is individually listed at Category A.

Can you guess which words are links? Answer below.

<details>
"rotunda" and "gatehouse" (not "bridge").
</details>

Also see the [actual article](https://en.wikipedia.org/wiki/Cullen_House#:~:text=within%20the%20grounds%20are%20a%20bridge%2C%20a%20rotunda%20and%20a%20gatehouse%2C%20each%20of%20which%20is%20individually%20listed%20at%20category%20a.). 

> I should also mention that I use only one label for all entities. Is it a big problem?

It's a small problem. It gives the model less information to work with; if you have multiple labels, their relationships can be hints to the model, and you don't have that. But if one label correctly describes your problem it's hard to improve it.

-----

I think you may just have a hard problem here, especially for recall, if you're actually trying to reproduce all Wikipedia internal links. It might be worth thinking about whether you can narrow your problem down a little bit.
